### PR TITLE
Upgrade Sacred to v0.8.1

### PR DESCRIPTION
**Pull request by shwang**
This PR updates our setup script to install ~~a patched version of Sacred v0.8.1 with two merged PRs.~~ Sacred v0.8.1.

We were unable to upgrade Sacred to the latest version because of issued noted in #109.  I recently submitted a patch to Sacred that makes ReadOnly{Dict,List} picklable, https://github.com/IDSIA/sacred/pull/737.

While running our Ray jobs with this patched version of Sacred, I also found that Sacred experiments run on Raylets fail to close their stdout-capturing `tee`s for some reason, which lead to an unhandled TimeoutError. This issue is fixed by https://github.com/IDSIA/sacred/pull/740, which warns that the tees failed to close instead of crashing.

~~I'm choosing to use our own version instead of IDSIA/sacred for now because the maintainers have not been responsive on issues and PRs for a while now, and I'd like to use `imitation` along with `realistic_benchmarks`, which requires a more recent version of `sacred`.~~

To avoid messing up PyPI and until the PRs are merged by IDSIA to the main repo we've decided to just address the two issues above by converting ReadOnlyContainer to regular python types when necessary, and by using `--capture=sys` to avoid the Tee TimeoutError.

### CODE REVIEW

**Comment by shwang** [comment]
In the end, I found it cleanest to used copy.deepcopy to "unfreeze" the `ReadOnlyDict`s in `parallel.py` anyways. Which is a shame considering that we could have foregone making ReadOnlyDict picklable if I figured this out in advance.

But https://github.com/IDSIA/sacred/pull/740 might have been necessary anyways.

**Comment by shwang** [comment]
Various fixes made to make the PR pass CI:

* Added code coverage for entrypoints to each Sacred script after refactoring deprecated Observer initializations. 
* Moved an assertion out of Sacred config and into the associated Sacred experiment command because the assertion was preventing `print_config` from running properly.

File: src/imitation/scripts/expert_demos.py [hunk 1]
```diff
@@ -121,6 +121,7 @@ def rollouts_and_policy(
 
         if init_tensorboard:
             sb_tensorboard_dir = osp.join(log_dir, "sb_tb")
+            init_rl_kwargs = dict(init_rl_kwargs)  # Cast sacred's ReadOnlyDict to dict.
```
Comment (AdamGleave) [inline src/imitation/scripts/expert_demos.py:124]: Is this necessary? Are we parallelizing things inside `expert_demos`?

File: src/imitation/scripts/parallel.py [hunk 2]
```diff
@@ -47,15 +49,28 @@ def parallel(
...
     # Explicitly set `data_dir` if parallelizing `train_adversarial`.
-    # We need this to automatically find rollout pickles Ray will automatically
-    # change the working directory for each Raylet.
+    # We need this to automatically find rollout pickles because Ray
+    # sets a new working directory for each Raylet.
     if sacred_ex_name == "train_adversarial":
         if "data_dir" not in base_config_updates:
-            base_config_updates["data_dir"] = os.path.join(os.getcwd(), "data/")
+            data_dir = os.path.join(os.getcwd(), "data/")
+            base_config_update = dict(base_config_updates)
```
Comment (AdamGleave) [inline src/imitation/scripts/parallel.py:66]: Typo? `base_config_update` != `base_config_updates`. If these are meant to be different then please rename to something more clearly distinct.

File: src/imitation/scripts/parallel.py [hunk 3]
```diff
@@ -47,15 +49,28 @@ def parallel(
...
+            base_config_update["data_dir"] = data_dir
 
     trainable = _ray_tune_sacred_wrapper(
-        sacred_ex_name, run_name, base_named_configs, base_config_updates
+        sacred_ex_name,
+        run_name,
+        copy.deepcopy(base_named_configs),
```
Comment (AdamGleave) [inline src/imitation/scripts/parallel.py:72]: Aren't `base_named_configs` just a list of strings? Can we just do `list(base_named_configs)`? I'm not sure I even understand why this is needed -- is `base_named_configs` a `FrozenList`? Good to add a comment explaining this.

File: src/imitation/scripts/parallel.py [hunk 4]
```diff
@@ -70,7 +85,7 @@ def parallel(
     try:
         ray.tune.run(
             trainable,
-            config=search_space,
+            config=copy.deepcopy(search_space),
```
Comment (AdamGleave) [inline src/imitation/scripts/parallel.py:88]: Again, a comment explaining this would be useful.

File: src/imitation/scripts/parallel.py [hunk 5]
```diff
@@ -115,16 +138,26 @@ def inner(config: dict, reporter) -> dict:
...
+        named_configs.extend(run_kwargs["named_configs"])
+        updated_run_kwargs["named_configs"] = named_configs
+
+        config_updates = {}
+        config_updates.update(base_config_updates)
+        config_updates.update(run_kwargs["config_updates"])
+        updated_run_kwargs["config_updates"] = config_updates
+
+        # Add other run_kwargs items to updated_run_kwargs.
+        for k, v in run_kwargs.items():
```
Comment (AdamGleave) [inline src/imitation/scripts/parallel.py:156]: To check my understanding: the semantics here are that `base_config_updates` should take priority over `run_kwargs` when both specify the same key?

File: src/imitation/scripts/train_adversarial.py [hunk 6]
```diff
@@ -108,25 +109,22 @@ def train(
...
         assert len(expert_trajs) >= n_expert_demos
         expert_trajs = expert_trajs[:n_expert_demos]
 
-    expert_stats = rollout.rollout_stats(expert_trajs)
-
     with networks.make_session():
         if init_tensorboard:
             sb_tensorboard_dir = osp.join(log_dir, "sb_tb")
-            kwargs = init_trainer_kwargs
+            kwargs = copy.deepcopy(init_trainer_kwargs)
```
Comment (AdamGleave) [inline src/imitation/scripts/train_adversarial.py:127]: Why is this here? I don't understand the need to have `copy.deepcopy` in script files (here or elsewhere). Non-pickleable objects will cause a problem before it even hits this line of code: they need to be addressed in the code that invokes Ray, not here.

File: tests/test_scripts.py [hunk 7]
```diff
@@ -6,24 +6,37 @@
...
+    expert_demos,
+    parallel,
+    train_adversarial,
+)
+
+ALL_SCRIPTS_MODS = [analyze, eval_policy, expert_demos, parallel, train_adversarial]
+
+
+@pytest.mark.parametrize("script_mod", ALL_SCRIPTS_MODS)
+def test_main_console(script_mod):
```
Comment (AdamGleave) [inline tests/test_scripts.py:31]: :+1:

File: setup.py [hunk 8]
```diff
@@ -55,9 +55,10 @@ def get_readme() -> str:
         "stable-baselines~=2.10.0",
         "jax~=0.1.66",
         "jaxlib~=0.1.47",
-        # sacred==0.7.5 build is broken without pymongo
-        # sacred>0.7.4 have non-picklable config objects (see GH #109)
-        "sacred==0.7.4",
+        # TODO(shwang): Install from IDSIA/sacred@master once
+        # https://github.com/IDSIA/sacred/pull/737 and
+        # https://github.com/IDSIA/sacred/pull/740 are merged.
+        "sacred @ git+https://github.com/HumanCompatibleAI/sacred@0.8.1-with-imit-compat",  # noqa: E501
```
Comment (AdamGleave) [inline setup.py:61]: We can't upload to PyPi if our requirements are not in PyPi, so I'm reluctant to make this change on a non-temporary basis.

File: src/imitation/scripts/parallel.py [hunk 9]
```diff
@@ -47,15 +49,28 @@ def parallel(
...
+            data_dir = os.path.join(os.getcwd(), "data/")
+            base_config_update = dict(base_config_updates)
+            base_config_update["data_dir"] = data_dir
 
     trainable = _ray_tune_sacred_wrapper(
-        sacred_ex_name, run_name, base_named_configs, base_config_updates
+        sacred_ex_name,
+        run_name,
+        copy.deepcopy(base_named_configs),
+        copy.deepcopy(base_config_updates),
```
Comment (AdamGleave) [inline src/imitation/scripts/parallel.py:73]: Again, a comment explaining this would be useful.

File: src/imitation/scripts/train_adversarial.py (hunk 6 above)
Comment (shwang) [reply src/imitation/scripts/train_adversarial.py:127]: The next line modifies `kwargs`, which is illegal without converting to `dict`. I chose to "defensively" deep copy in case for some reason in the future any code downstream (including in the imitation learning algorithm) wants to modify any dicts or lists inside the dict.

File: src/imitation/scripts/parallel.py (hunk 2 above)
Comment (shwang) [reply src/imitation/scripts/parallel.py:66]: :+1: this is a typo.

File: src/imitation/scripts/parallel.py (hunk 5 above)
Comment (shwang) [reply src/imitation/scripts/parallel.py:156]: No, it's actually the opposite. Because `base_config_updates` gets `update`-applied first, and then `run_kwargs["named_configs"]` overwrites this with the second `update`.

Looking back at the docstring, this doesn't seem well explained.
- [ ] Better docs for base_*_config

File: src/imitation/scripts/expert_demos.py (hunk 1 above)
Comment (shwang) [reply src/imitation/scripts/expert_demos.py:124]: Need this in order to modify `tensorboard_log` in the next line.

File: src/imitation/scripts/parallel.py (hunk 3 above)
Comment (shwang) [reply src/imitation/scripts/parallel.py:72]: I added in the `deepcopy` to check if non-forked Sacred would work, and left it in for a confused reason that probably isn't worth elaborating on.

non-forked Sacred didn't work because it would deterministically fail on the issue fixed by https://github.com/IDSIA/sacred/pull/740.

File: src/imitation/scripts/parallel.py (hunk 4 above)
Comment (shwang) [reply src/imitation/scripts/parallel.py:88]: (see explanation above in https://github.com/HumanCompatibleAI/imitation/pull/194#discussion_r434180592)

File: src/imitation/scripts/parallel.py (hunk 9 above)
Comment (shwang) [reply src/imitation/scripts/parallel.py:73]: (See https://github.com/HumanCompatibleAI/imitation/pull/194#discussion_r434180592)

Comment (shwang) [reply src/imitation/scripts/parallel.py:73]: Looking back `deepcopy` + ReadOnlyDict is pretty esoteric knowledge. Agreed that this would have done better with a comment at least explaining what `deepcopy` does here

File: setup.py (hunk 8 above)
Comment (shwang) [reply setup.py:61]: This is a reasonable objection, I forgot that PyPI has this requirement

**Comment by shwang** [comment]
@AdamGleave thanks for your thorough review!

Especially with the PyPI consideration in mind I agree that it isn't a good idea to point to the Sacred fork as a requirement.

I think I'll do the thing you suggested where I install a personal `imitation` branch when I'm in a python environment that also needs the latest version of Sacred.

**Comment by AdamGleave** [comment]
> I think I'll do the thing you suggested where I install a personal `imitation` branch when I'm in a python environment that also needs the latest version of Sacred.

I would like to upgrade to Sacred 0.8.1 still...

With `deepcopy`, is the only blocking issue the `tee` bug? I'm more optimistic about some kind of fix there making it's way into Sacred eventually (whereas the pickle issue I think is more a difference of opinion between us and the maintainers). Is there a way we can workaround it? It seems to be very dependent on platform, so if it doesn't break unit tests on CI, we can maybe just add a warning in the README with a link to your fork...

**Comment by shwang** [comment]
Oops, I missed your most recent comment! Looking into CI+Sacred 0.8.1 stuff.

**Comment by shwang** [comment]
Tried it out, and in the first CI run one of the parallel_ex test cases and a expert_demos_ex case failed on the Tee error.

https://app.circleci.com/pipelines/github/HumanCompatibleAI/imitation/402/workflows/33b80186-3b6c-4594-82b3-76071abab6da/jobs/836/parallel-runs/2/steps/2-107

**Comment by AdamGleave** [comment]
> Tried it out, and in the first CI run one of the parallel_ex test cases and a expert_demos_ex case failed on the Tee error.
> 
> https://app.circleci.com/pipelines/github/HumanCompatibleAI/imitation/402/workflows/33b80186-3b6c-4594-82b3-76071abab6da/jobs/836/parallel-runs/2/steps/2-107

I think this can probably be worked around by changing the [capture mode](https://sacred.readthedocs.io/en/stable/experiment.html#capturing-stdout-stderr). The nuclear option is to just set capture mode to `no`, but I think setting it to `sys` would also work. It'll miss some print statements from libraries etc but I don't think we really need these? Certainly not for tests.

I'd ideally still like to fix the root cause, but I think Sacred maintainers would be receptive to any PRs on the `tee` issue we might come up with.

**Comment by shwang** [comment]
The `sys` capture mode sounds appropriate for tests. This could be a fixture in the test_scripts.py file, and we could add a little comment that the fixture can be removed for more detailed error logging and stuff.

**Comment by shwang** [comment]
"sys" seems to fix things on my end, at least with the manual tests I"ve managed so far, setting `-C sys` from the command line.

Unfortunately, AFAIK using `SETTINGS["CAPTURE_MODE"] = ...` is a no-op because this setting takes 2nd priority over the command line option `-C`, which is by default set to `"fd"` on Unix systems.

**Comment by shwang** [comment]
Looking into whether this is a workaround for this...

**Comment by shwang** [comment]
https://github.com/IDSIA/sacred/issues/742

File: src/imitation/data/rollout.py [hunk 10]
```diff
@@ -345,8 +345,11 @@ def rollout_stats(trajectories: Sequence[types.TrajectoryWithRew]) -> dict:
     stat_names = ["min", "mean", "std", "max"]
     for desc_name, desc_vals in traj_descriptors.items():
         for stat_name in stat_names:
-            stat_value = getattr(np, stat_name)(desc_vals)
-            out_stats[f"{desc_name}_{stat_name}"] = stat_value
+            stat_value: np.generic = getattr(np, stat_name)(desc_vals)
+            out_stats[f"{desc_name}_{stat_name}"] = stat_value.item()
+
+    for v in out_stats.values():
+        assert isinstance(v, (int, float))
```
Comment (shwang) [inline src/imitation/data/rollout.py:352]: These return values need to be non-numpy because otherwise they don't return properly as run-results from expert_demos_ex, etc.

I could add this add as a comment to the code itself if that seems helpful.

File: src/imitation/scripts/expert_demos.py (hunk 1 above)
Comment (shwang) [reply src/imitation/scripts/expert_demos.py:124]: Modifying the comment to be explicit why we are ~"casting" to dict here.

File: src/imitation/scripts/parallel.py (hunk 5 above)
Comment (shwang) [reply src/imitation/scripts/parallel.py:164]: Added better docs.

File: setup.py [hunk 11]
```diff
@@ -55,9 +55,7 @@ def get_readme() -> str:
         "stable-baselines~=2.10.0",
         "jax~=0.1.66",
         "jaxlib~=0.1.47",
-        # sacred==0.7.5 build is broken without pymongo
-        # sacred>0.7.4 have non-picklable config objects (see GH #109)
-        "sacred==0.7.4",
+        "sacred",
```
Comment (shwang) [inline setup.py:58]: Should we lock to 0.8.1?

File: src/imitation/scripts/parallel.py (hunk 5 above)
Comment (shwang) [reply src/imitation/scripts/parallel.py:178]: This for loop adds all keys except "config_updates" and "named_configs" from the search_space into the run_kwargs. Hopefully the docs should clarify this a bit.

It's pretty messy getting the base_config_updates + config_updates in through the ray.tune API.

File: setup.py (hunk 11 above)
Comment (AdamGleave) [reply setup.py]: I'd rather not -- you're then forcing users of `imitation` to choose a particular `sacred` version. You probably want to have a minimum version though?

File: src/imitation/data/rollout.py (hunk 10 above)
Comment (AdamGleave) [reply src/imitation/data/rollout.py:352]: Hmm, it might be worth mentioning briefly that the NumPy is returning a NumPy type even for scalars, but I don't think we need to mention the scripts since this is in a non-script file.

File: src/imitation/scripts/parallel.py [hunk 12]
```diff
@@ -104,6 +137,18 @@ def _ray_tune_sacred_wrapper(
     """
 
     def inner(config: dict, reporter) -> dict:
+        """Trainable function with the correct signature for `ray.tune`.
+
+        Args:
+            config: Keyword arguments for `ex.run()`, where `ex` is the
+                `sacred.Experiment` instance associated with `sacred_ex_name`.
+        """
+        # Set Sacred capture mode to "sys" because default "fd" option leads to error.
```
Comment (AdamGleave) [inline src/imitation/scripts/parallel.py:146]: I'd make this a TODO or SOMEDAY since hopefully one day the Sacred issue is resolved and we can remove this workaround

File: tests/test_scripts.py [hunk 13]
```diff
@@ -6,24 +6,50 @@
...
+    expert_demos,
+    parallel,
+    train_adversarial,
+)
+
+ALL_SCRIPTS_MODS = [analyze, eval_policy, expert_demos, parallel, train_adversarial]
+
+
+@pytest.fixture(autouse=True)
+def sacred_capture_use_sys():
```
Comment (AdamGleave) [inline tests/test_scripts.py:32]: Should probably also have a TODO/SOMEDAY on it since a workaround for a bug in Sacred.

File: tests/test_scripts.py [hunk 14]
```diff
@@ -6,24 +6,50 @@
...
+
+    See https://github.com/IDSIA/sacred/issues/289."""
+    temp = sacred.SETTINGS["CAPTURE_MODE"]
+    sacred.SETTINGS.CAPTURE_MODE = "sys"
+    yield
+    sacred.SETTINGS.CAPTURE_MODE = temp
+
+
+@pytest.mark.parametrize("script_mod", ALL_SCRIPTS_MODS)
+def test_main_console(script_mod):
```
Comment (AdamGleave) [inline tests/test_scripts.py:43]: Should we remove the `# pragma: no cover` lines then from the main console blocks?

File: src/imitation/scripts/parallel.py [hunk 15]
```diff
@@ -27,35 +30,65 @@ def parallel(
...
+            CLI '--name' option on the inner experiment. Offline analysis jobs can use
+            this argument to group similar data.
+        search_space: A dictionary which can contain `ray.tune.grid_search` and is
+            passed as the `config` argument to `ray.tune.run()`. After the
+            `search_space` is transformed by Ray, it passed into
+            `sacred_ex.run(**run_kwargs)` as `run_kwargs` (`sacred_ex` is the Sacred
+            Experiment selected via `sacred_ex_name`). Usually `search_space` only has
+            the keys "named_configs" and "config_updates", but any parameter names
+            to `sacred.Experiment.run()` are okay.
+        base_named_configs: Default Sacred Run named configs. Any named configs
```
[the remark this answers is not in the document]
Comment (shwang) [reply src/imitation/scripts/parallel.py:48]: probably meant to just write "named configs" there.

File: tests/test_scripts.py (hunk 14 above)
Comment (shwang) [reply tests/test_scripts.py:43]: `main_console` never had `pragma: no cover`. Did you mean the `if __name__ == "__main__"` blocks?

`test_main_console` doesn't account for that, as it calls `main_console()` directly.

File: src/imitation/scripts/parallel.py [hunk 16]
```diff
@@ -27,35 +30,65 @@ def parallel(
...
-      upload_dir: `upload_dir` argument to `ray.tune.run()`.
+        sacred_ex_name: The Sacred experiment to tune. Either "expert_demos" or
+            "train_adversarial".
+        run_name: A name describing this parallelizing experiment.
+            This argument is also passed to `ray.tune.run` as the `name` argument.
+            It is also saved in 'sacred/run.json' of each inner Sacred experiment
+            under the 'experiment.name' key. This is equivalent to using the Sacred
+            CLI '--name' option on the inner experiment. Offline analysis jobs can use
+            this argument to group similar data.
+        search_space: A dictionary which can contain `ray.tune.grid_search` and is
```
[the remark this answers is not in the document]
Comment (shwang) [reply src/imitation/scripts/parallel.py:41]: Updated comment

File: tests/test_scripts.py (hunk 14 above)
Comment (shwang) [reply tests/test_scripts.py:43]: If coverage for the __main__ block is something we want to have, I can add that test in a separate PR.

File: src/imitation/data/rollout.py (hunk 10 above)
Comment (shwang) [reply src/imitation/data/rollout.py:352]: Added a comment to that effect.
